src/game/common/compression/compressionmanager.cpp

```diff
@@ -124,26 +196,80 @@ int CompressionManager::Get_Uncompressed_Size(const void *data, int size)
         return size;
     }
 
-    return le32toh(static_cast<const int32_t *>(data)[1]);
+    const ComprHeader *header = static_cast<const ComprHeader *>(data);
+    return le32toh(header->uncomp_size);
 }
 
 /**
- * @brief Decompress possibly compressed data. Only handles RefPack compression.
+ * @brief Compress uncompressed data. Only handles RefPack and ZLib compression.
  */
-int CompressionManager::Decompress_Data(void *src, int src_size, void *dst, int dst_size)
+int CompressionManager::Compress_Data(CompressionType type, void *src, int src_size, void *dst, int dst_size)
 {
-    if (src_size <= 7) {
+    if (dst_size < sizeof(ComprHeader)) {
         return 0;
     }
 
-    switch (Get_Compression_Type(src, src_size)) {
-        case COMPRESSION_EAR: // RefPack
-            src_size -= 8;
-            return RefPack_Uncompress(dst, static_cast<const uint8_t *>(src) + 8, &src_size);
+    uint32_t fourcc = Get_Compression_FourCC(type);
+
+    // Initialize the header. We could think about not writing the fourcc when format is not supported,
+    // but probably doesn't really matter
+    ComprHeader *header = static_cast<ComprHeader *>(dst);
+    memcpy(header->fourcc, &fourcc, sizeof(uint32_t));
+    header->uncomp_size = 0;
 
-        // Original game handles all these formats, ZH only appears to use RefPack however.
+    uint32_t compr_size = 0;
+    uint8_t *dst_data = static_cast<uint8_t *>(dst) + sizeof(ComprHeader);
+
+    switch (type) {
+        case COMPRESSION_EAR:
+            compr_size = RefPack_Compress(dst_data, src, src_size, false);
+            if (compr_size > 0) {
+                header->uncomp_size = src_size;
+                return compr_size + sizeof(ComprHeader);
+            }
+            break;
+        case COMPRESSION_ZL1:
+        case COMPRESSION_ZL2:
+        case COMPRESSION_ZL3:
+        case COMPRESSION_ZL4:
+        case COMPRESSION_ZL5:
+        case COMPRESSION_ZL6:
+        case COMPRESSION_ZL7:
+        case COMPRESSION_ZL8:
+        case COMPRESSION_ZL9:
+#if BUILD_WITH_ZLIB
+            compr_size = Zlib_Compress(dst_data, dst_size - sizeof(ComprHeader), src, src_size, type + 1 - COMPRESSION_ZL1);
+            if (compr_size > 0) {
+                header->uncomp_size = src_size;
+                return compr_size + sizeof(ComprHeader);
+            }
```

Comment:
Are we not missing a break here in case Zlib_Compress returns < 0?

Reply:
It can't return anything lower than 0